### PR TITLE
Support externref in EM_JS functions with dynamic linking

### PR DESCRIPTION
**Pull request by hoodmane**

### CODE REVIEW

**Comment by sbc100** [comment]
Nice!  Looks goo to me if the test pass

File: test/code_size/hello_webgl_wasm2js.json [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
-  "a.js": 17454,
-  "a.js.gz": 7873,
+  "a.js": 17450,
+  "a.js.gz": 7869,
   "a.mem": 3171,
   "a.mem.gz": 2713,
-  "total": 21192,
-  "total_gz": 10965
+  "total": 21188,
+  "total_gz": 10961
```
Comment (sbc100) [inline test/code_size/hello_webgl_wasm2js.json:9]: Are these code size changes actually due to this change?  If not I would not include them.

Comment (hoodmane) [reply test/code_size/hello_webgl_wasm2js.json]: I am not sure. I am just experimenting trying to get the tests to pass.

Comment (hoodmane) [reply test/code_size/hello_webgl_wasm2js.json]: I see:
```
test_minimal_runtime_code_size_math (test_other.other) ... FAIL                    3512 != 3511
test_minimal_runtime_code_size_random_printf_wasm (test_other.other) ... FAIL     12766 != 12774
test_minimal_runtime_code_size_random_printf_wasm2js (test_other.other) ... FAIL  17344 != 17351
```

Comment (sbc100) [reply test/code_size/hello_webgl_wasm2js.json]: Do you see those failures on main, or only your branch?

File: test/test_core.py [hunk 2]
```diff
@@ -9749,6 +9749,19 @@ def test_externref(self):
     self.emcc_args += ['-mreference-types']
     self.do_core_test('test_externref.c', libraries=['asm.o'])
 
+  @parameterized({
+    '': [False],
+    'dynlink': [True]
+  })
+  @requires_node
+  @no_wasm2js('wasm2js does not support reference types')
+  def test_externref2(self, dynlink):
```
Comment (sbc100) [inline test/test_core.py:9758]: Can we give this test a better name.  How about `test_externref_em_js` ?  (and name the source file likewise).

File: test/core/test_externref2.c [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#include "emscripten.h"
```
Comment (sbc100) [inline test/core/test_externref2.c:1]: `#include <emscripten/em_js.h>`

File: test/code_size/hello_webgl_wasm2js.json (hunk 1 above)
Comment (hoodmane) [reply test/code_size/hello_webgl_wasm2js.json]: I think locally they fail in both cases. But the failures I indicated above are CI failures caused by this patch.

File: test/test_core.py (hunk 2 above)
Comment (hoodmane) [reply test/test_core.py]: done.

File: test/code_size/hello_webgl_wasm2js.json (hunk 1 above)
Comment (sbc100) [reply test/code_size/hello_webgl_wasm2js.json]: In order to keep get accurate code size number locally I recommend using `tot` build of emsdk (if you use emsdk).     If you don't use emsdk then instead you will want to make sure you have llvm and binaryen both built from `main`.   You will also want to do `./emcc --clear-cache` before running the test.

I normally run this command: `./test/runner other.*code_size* other.*metadce*  --rebase`.    You the files that are checked in are up-to-date and you have a tot build of llvm and binaryen that command should not generate any changes.

Comment (sbc100) [reply test/code_size/hello_webgl_wasm2js.json:9]: Hmm .. now it looks like these change improves code size which looks very odd give that it only adds code. 

Let me download this PR and see if actually affects any of these code size test..

Comment (sbc100) [reply test/code_size/hello_webgl_wasm2js.json:9]: I verfied that this change only seems to effect `test_metadce_hello_dylink` and even then well with the tolerance.  

So I think can you safely revert all these expectation files.

Comment (hoodmane) [reply test/code_size/hello_webgl_wasm2js.json]: Okay reverted it. I still don't understand the test failures.

**Comment by hoodmane** [comment]
Okay the tests all pass now, so all the previous failures were just flakes. I guess that's not too surprising.

**Review by sbc100 — APPROVED** [review]

**Pull request merged**